File: chronicle-pc/pipeline/chronicle_pipeline/markdown_index.py
"""Live ``_system/index.md`` agent shortlist (PARA + journal day files).

Sqlite under ``index/`` remains the RAG source of truth. This markdown file is a
regenerated skim index for Claude Code / Obsidian — not hand-edited SoT.

Rebuild via ``chronicle rebuild-markdown-index``, ``chronicle index --write-markdown``,
or ``POST /vault/rebuild-index``. Inbox filing and ``_staging`` import stay
skill/CLI-driven (not SPA wizards).
"""

from __future__ import annotations

import logging
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from . import path_map
from .paths import atomic_write_text, resolve_chronicle_dir
from .vault_paths import JOURNAL_DIR

log = logging.getLogger("chronicle.markdown_index")

INDEX_REL = "_system/index.md"

_FRONTMATTER_RE = re.compile(r"\A---\r?\n(.*?)\r?\n---\r?\n?", re.DOTALL)
_H1_RE = re.compile(r"^#\s+(.+)$", re.MULTILINE)

INDEX_HEADER = """# Vault index

Agent shortlist (regenerated — do not hand-edit as SoT). Sqlite RAG lives under
`index/`. Rebuild: `chronicle rebuild-markdown-index` or `POST /vault/rebuild-index`.

Format: `title | type | tags | updated`

"""


def _parse_frontmatter(text: str) -> dict[str, str]:
    m = _FRONTMATTER_RE.match(text)
    if not m:
        return {}
    meta: dict[str, str] = {}
    for line in m.group(1).splitlines():
        if ":" not in line:
            continue
        key, _, val = line.partition(":")
        meta[key.strip().lower()] = val.strip().strip("\"'")
    return meta


def _format_tags(raw: str) -> str:
    raw = (raw or "").strip()
    if not raw:
        return ""
    if raw.startswith("[") and raw.endswith("]"):
        inner = raw[1:-1].strip()
        if not inner:
            return ""
        parts = [p.strip().strip("\"'") for p in inner.split(",")]
        return ", ".join(p for p in parts if p)
    return raw


def _file_mtime_date(path: Path) -> str:
    try:
        ts = path.stat().st_mtime
        return datetime.fromtimestamp(ts, tz=timezone.utc).astimezone().date().isoformat()
    except OSError:
        return datetime.now(timezone.utc).astimezone().date().isoformat()


def _row_for_file(
    rel: str,
    path: Path,
    text: str,
    *,
    default_type: str,
) -> dict[str, str]:
    meta = _parse_frontmatter(text)
    title = (meta.get("title") or "").strip()
    if not title:
        h1 = _H1_RE.search(text)
        title = h1.group(1).strip() if h1 else Path(rel).stem
    note_type = (meta.get("type") or "").strip() or default_type
    tags = _format_tags(meta.get("tags") or "")
    updated = (meta.get("updated") or meta.get("created") or "").strip()
    if not updated:
        updated = _file_mtime_date(path)
    return {
        "rel": rel,
        "title": title,
        "type": note_type,
        "tags": tags,
        "updated": updated,
    }


def collect_index_rows(root: Path) -> list[dict[str, str]]:
    """Walk PARA knowledge notes + ``40-Journal/`` day files."""
    rows: list[dict[str, str]] = []
    seen: set[str] = set()

    for rel, path in path_map.iter_knowledge_md(root):
        if rel in seen:
            continue
        try:
            text = path.read_text(encoding="utf-8", errors="replace")
        except OSError:
            continue
        seen.add(rel)
        rows.append(_row_for_file(rel, path, text, default_type="note"))

    journal_root = root / JOURNAL_DIR
    if journal_root.is_dir():
        for path in sorted(journal_root.rglob("*.md")):
            if path.name.startswith(".") or ".sync-conflict" in path.name:
                continue
            if path_map.is_chrome_basename(path.name):
                continue
            rel = str(path.relative_to(root)).replace("\\", "/")
            if rel in seen:
                continue
            try:
                text = path.read_text(encoding="utf-8", errors="replace")
            except OSError:
                continue
            seen.add(rel)
            rows.append(_row_for_file(rel, path, text, default_type="journal"))

    rows.sort(key=lambda r: (r["updated"], r["title"].lower()), reverse=True)
    return rows


def format_index_markdown(rows: list[dict[str, str]]) -> str:
    lines = [INDEX_HEADER.rstrip(), ""]
    for row in rows:
        tags = row["tags"] or "—"
        lines.append(
            f"{row['title']} | {row['type']} | {tags} | {row['updated']}"
        )
    lines.append("")
    return "\n".join(lines)


def rebuild_markdown_index(
    chronicle_dir: Path | str | None = None,
    *,
    dry_run: bool = False,
) -> dict[str, Any]:
    """Rewrite ``_system/index.md`` from PARA + journal day files."""
    root = resolve_chronicle_dir(chronicle_dir)
    rows = collect_index_rows(root)
    content = format_index_markdown(rows)
    dest = root / INDEX_REL
    if dry_run:
        log.info(
            "[dry-run] would write %s (%d rows, %d bytes)",
            INDEX_REL,
            len(rows),
            len(content.encode("utf-8")),
        )
        return {
            "path": INDEX_REL,
            "rows": len(rows),
            "dry_run": True,
            "would_write": True,
        }

    dest.parent.mkdir(parents=True, exist_ok=True)
    atomic_write_text(dest, content if content.endswith("\n") else content + "\n")
    log.info("Wrote %s (%d rows)", INDEX_REL, len(rows))
    return {
        "path": INDEX_REL,
        "rows": len(rows),
        "dry_run": False,
        "ok": True,
    }

File: chronicle-pc/server/src/markdown_index.rs
//! _system/index.md agent shortlist (markdown_index.py).

use std::path::Path;

use serde_json::{json, Value};

use crate::errors::ChronicleError;
use crate::frontmatter::parse_frontmatter;
use crate::path_map;
use crate::paths::atomic_write_text;

pub const INDEX_REL: &str = "_system/index.md";

const HEADER: &str = r"# Vault index

Agent shortlist (regenerated — do not hand-edit as SoT). Sqlite RAG lives under
`index/`. Rebuild: `chronicle rebuild-markdown-index` or `POST /vault/rebuild-index`.

Format: `title | type | tags | updated`";

struct Row {
    title: String,
    kind: String,
    tags: String,
    updated: String,
}

fn row_for_file(path: &std::path::Path, root: &Path, default_type: &str) -> Option<Row> {
    let bytes = std::fs::read(path).ok()?;
    let text = String::from_utf8_lossy(&bytes).to_string();
    let parsed = parse_frontmatter(&text);
    let fm = &parsed.fm;

    let get = |k: &str| fm.get(k).cloned();

    let title = get("title")
        .filter(|t| !t.is_empty())
        .or_else(|| {
            regex::Regex::new(r"(?m)^#\s+(.+)$")
                .ok()?
                .captures(&text)
                .and_then(|c| c.get(1))
                .map(|m| m.as_str().trim().to_string())
        })
        .unwrap_or_else(|| {
            path.file_stem()
                .map(|s| s.to_string_lossy().to_string())
                .unwrap_or_default()
        });

    let kind = get("type").filter(|t| !t.is_empty()).unwrap_or_else(|| default_type.to_string());

    let tags_raw = get("tags").unwrap_or_default();
    let tags = if tags_raw.starts_with('[') && tags_raw.ends_with(']') {
        tags_raw
            .trim_start_matches('[')
            .trim_end_matches(']')
            .split(',')
            .map(|p| p.trim().trim_matches('"').trim_matches('\'').to_string())
            .filter(|p| !p.is_empty())
            .collect::<Vec<_>>()
            .join(", ")
    } else {
        tags_raw.clone()
    };

    let updated = get("updated")
        .filter(|u| !u.is_empty())
        .or_else(|| get("created"))
        .unwrap_or_else(|| {
            // File mtime as local date ISO.
            std::fs::metadata(path)
                .and_then(|m| m.modified())
                .ok()
                .and_then(|t| {
                    let dt: chrono::DateTime<chrono::Utc> = t.into();
                    Some(dt.with_timezone(&chrono::Local).format("%Y-%m-%d").to_string())
                })
                .unwrap_or_else(|| chrono::Local::now().format("%Y-%m-%d").to_string())
        });
    let _ = root;
    Some(Row { title, kind, tags, updated })
}

fn collect_rows(root: &Path) -> Vec<Row> {
    let mut rows: Vec<Row> = Vec::new();
    let mut seen: std::collections::HashSet<String> = Default::default();

    for (rel, path) in path_map::iter_knowledge_md(root) {
        if seen.insert(rel) {
            if let Some(r) = row_for_file(&path, root, "note") {
                rows.push(r);
            }
        }
    }
    let journal_root = root.join("40-Journal");
    if journal_root.is_dir() {
        let mut paths: Vec<std::path::PathBuf> = walk_md(&journal_root);
        paths.sort();
        for p in paths {
            let name = p.file_name().map(|n| n.to_string_lossy().to_string()).unwrap_or_default();
            if name.starts_with('.') || name.contains(".sync-conflict") {
                continue;
            }
            let rel = p
                .strip_prefix(root)
                .map(|r| r.to_string_lossy().replace('\\', "/"))
                .unwrap_or_default();
            if seen.insert(rel) {
                if let Some(r) = row_for_file(&p, root, "journal") {
                    rows.push(r);
                }
            }
        }
    }
    rows.sort_by(|a, b| (&b.updated, b.title.to_lowercase()).cmp(&(&a.updated, a.title.to_lowercase())));
    rows
}

fn walk_md(dir: &Path) -> Vec<std::path::PathBuf> {
    let mut out = Vec::new();
    crate::paths::walk_files_filtered(dir, &mut out, 0, &|_p, name| name.ends_with(".md"));
    out
}

pub fn rebuild_markdown_index(root: &Path, dry_run: bool) -> Result<Value, ChronicleError> {
    let rows = collect_rows(root);
    let mut body = String::from(HEADER.trim_end());
    body.push_str("\n\n");
    for r in &rows {
        let tag_field = if r.tags.is_empty() { "—" } else { r.tags.as_str() };
        body.push_str(&format!("{} | {} | {} | {}\n", r.title, r.kind, tag_field, r.updated));
    }

    let path = root.join(INDEX_REL);
    if dry_run {
        return Ok(json!({
            "path": INDEX_REL,
            "rows": rows.len(),
            "dry_run": true,
            "would_write": true,
        }));
    }
    atomic_write_text(&path, &body)?;
    Ok(json!({
        "path": INDEX_REL,
        "rows": rows.len(),
        "dry_run": false,
        "ok": true,
    }))
}
